### DISCUSSION TITLE
"max user processes" limit is not set correctly in terminal

### DISCUSSION THREAD

**Discussion by OrenMustaki** [coder/code-server#5393] [Q&A] [answered]
Hi

When I'm opening terminal in code server , CTRL+SHIFT+~, the value of "max user processes" ( ulimit -u ) is set to 127582.
From the same terminal, if I open a new shell, or "ssh localhost", I get a correct value of 1042.
The value I get if open a terminal by other means is 1024 which is correct and set using /etc/limit.d/.
I checked it on a few code-server and the behavior is the same

What am I missing ?

Oren

**Comment by jsjoeio** [comment]
Hmm... @code-asher any ideas? Is this something we're manually setting somewhere?

**Reply by code-asher** [reply]
I do not think we are setting this anywhere.  If we had a way to reliably reproduce we could try debugging it.

**Reply by OrenMustaki** [reply]
To reproduce the issue, as root create a file /etc/security/limits.d/test.conf

`* hard nproc 512`
`* soft nproc 512`

open a new shell as a non root user and see the value using :
`uname -u`
it should be 512
open code-server and open terminal
CTRL + SHIFT + ~
run 
`uname -u`

**Reply by OrenMustaki** [reply]
did anyone reproduce it or is it only on my systems ?

**Reply by code-asher** [reply]
I tried these steps but when I run `ulimit -u` it always says `unlimited`.  I am running inside a Docker container on the Coder platform so that might be why.

**Reply by jsjoeio** [reply]
I'm testing now on macOS

**Reply by jsjoeio** [reply]
hmm i'm getting 2784 when running `ulimit -u`

<img width="873" alt="image" src="https://user-images.githubusercontent.com/3806031/183519172-2c1eeeb7-be58-4aea-98ab-0293a6a9f4fc.png">

are you sure this isn't specific to your system?

**Reply by OrenMustaki** [reply]
Iv'e tried it on few hosts and its the same.
the numbers might be different but the limit does not
seems to be enforced. if i reconnect using ssh the value is correct.

<img width="481" alt="image" src="https://user-images.githubusercontent.com/38783312/183577555-af0b95ea-aa8b-4613-b879-44d629309d1f.png">

![image](https://user-images.githubusercontent.com/38783312/183577787-b32bc80c-4af4-42f7-8ac2-d43e5d039c91.png)

**Reply by jsjoeio** [reply]
ah i see. hmm...to be quite honest, i'm not sure what else i can do to help since i'm not sure how this happening.

you could try reproducing with [Visual Studio Code Server](https://code.visualstudio.com/docs/remote/vscode-server)

**Reply by OrenMustaki** [reply]
I tried it with Visual Studio Code Server. it works as expected.
![image](https://user-images.githubusercontent.com/38783312/183813953-01b4c7d5-95fb-4f6e-a8d8-7bbde2fe4048.png)

for some reason my shell under code-server is invoked without any limit. probably it inherit the limits
from the invoking shell.
it can wreak havoc on a system if someone run a fork bomb by mistake or intentionally

**Reply by jsjoeio** [reply]
Hmm...strange! Well thanks for trying. Unfortunately, since we cannot successfully reproduce I'm not sure there is much we can do to help. I'm sorry about that

**Reply by code-asher** [reply]
Inherited from the invoking shell makes sense to me.  I imagine this could depend on how you are starting code-server; is it systemd or something?  Did you start Visual Studio Code Server the same way?

**Reply by OrenMustaki** [reply]
Yes, im running it with systemd.
With the Visual Studio Code im using Remote SSH extension to connect .
Ill try to look into the systemd service. it might have something to do with it.

**Reply by OrenMustaki** [reply]
that was it !!!
Since daemons started by systemd don't employ pam login session, the limits can be set only in the service unit file.
I had to add LimitNPROC=XXX  to the code-server systemd file under [Service]

Thanks

Oren

**Reply by jsjoeio** [reply]
@code-asher does it again.

**Comment by jsjoeio** [comment, marked as answer]
> Since daemons started by systemd don't employ pam login session, the limits can be set only in the service unit file.
I had to add LimitNPROC=XXX to the code-server systemd file under [Service]

[Link](https://github.com/coder/code-server/discussions/5393#discussioncomment-3369379)

Solved by: @code-asher
